### PR TITLE
Event thrift encoding and refactoring dataInterfaces

### PR DESCRIPTION
**Pull request by longquanzheng**
We want to support binary encoding for our history events. While doing that, we decided to refactor our dataInterface so that our application layer doesn't need to know about encoding/decoding. Also our persistence layer is broken down into two, the middle layer will be responsible for decoding/encoding and calculating size information.

### CODE REVIEW

File: common/persistence/dataStoreImpl.go [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+	"github.com/uber-common/bark"
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/logging"
+)
+
+type (
+
+	// executionManagerImpl implements ExecutionManager based on PersistenceExecutionManager, statsComputer and HistorySerializer
+	executionManagerImpl struct {
```
Comment (samarabbas) [inline common/persistence/dataStoreImpl.go]: Can you separate out each one of manager implementations in its own file.

File: common/persistence/statsComputer.go [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+	// statsComputer is to computing struct sizes after serialization
+	statsComputer struct{}
+)
+
+func (sc *statsComputer) computeMutableStateStats(req *PersistenceUpdateWorkflowExecutionRequest) *MutableStateStats {
+	executionInfoSize := computeExecutionInfoSize(req.ExecutionInfo)
+
+	activityInfoCount := 0
+	activityInfoSize := 0
+	for _, ai := range req.UpsertActivityInfos {
```
Comment (samarabbas) [inline common/persistence/statsComputer.go]: This is not going to work.  Mutable state stats cannot be computed using update request.  It only has the upsert activity/timer/child/signal infos, not the entire size of mutable state.

File: common/persistence/dataStoreImpl.go (hunk 1 above)
Comment (longquanzheng) [reply common/persistence/dataStoreImpl.go]: Got it

File: common/persistence/historySerializer.go [hunk 3]
```diff
@@ -23,25 +23,23 @@ package persistence
...
 type (
-	// HistorySerializer is used to serialize/deserialize history
+	// HistorySerializer is used by persistence to serialize/deserialize history event(s)
+	// It will only be used inside persistence, so that serialize/deserialize is transparent for application
 	HistorySerializer interface {
-		Serialize(batch *HistoryEventBatch) (*SerializedHistoryEventBatch, error)
-		Deserialize(batch *SerializedHistoryEventBatch) (*HistoryEventBatch, error)
-	}
+		// serialize/deserialize history events
+		SerializeBatchEvents(batch *workflow.History, encodingType common.EncodingType) (*DataBlob, error)
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/historySerializer.go]: We can do that. I thought you only want that flatening in DataInterface.

File: common/persistence/persistenceInterface.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+	// PersistenceShardManager is a lower level of ShardManager
+	PersistenceShardManager = ShardManager
+	// PersistenceTaskManager is a lower level of TaskManager
+	PersistenceTaskManager = TaskManager
+	// PersistenceMetadataManager is a lower level of MetadataManager
+	PersistenceMetadataManager = MetadataManager
+
+	// PersistenceExecutionManager is used to manage workflow executions for Persistence layer
+	PersistenceExecutionManager interface {
```
Comment (samarabbas) [inline common/persistence/persistenceInterface.go]: Let's use Store naming convention here.
ExecutionStore
EventsStore
etc

Comment (longquanzheng) [reply common/persistence/persistenceInterface.go]: OK sounds good to me.
BTW EventsStore I think you meant HistoryStore

File: common/persistence/historySerializer.go [hunk 5]
```diff
@@ -63,79 +61,120 @@ type (
...
-// of HistorySerializerFactory
-func NewHistorySerializerFactory() HistorySerializerFactory {
-	return &serializerFactoryImpl{
-		jsonSerializer: NewJSONHistorySerializer(),
+func (t *serializerImpl) SerializeEvent(event *workflow.HistoryEvent, encodingType common.EncodingType) (*DataBlob, error) {
+	if event == nil {
+		event = &workflow.HistoryEvent{}
+	}
+	switch encodingType {
+	case common.EncodingTypeGob:
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/historySerializer.go]: We need to return error here to make unit test happy. The default case is Json which shouldn't return error.

**Pull request merged**